### PR TITLE
[Internal] FaultInjection: Adds support for FaultInjection for RNTBD V3 SDK

### LINKED ISSUE #4256: Add Fault Injection support Interfaces and Abstract Classes to SDK
Add fault injection support for the V3 .NET SDK. This will allow for adding an implementation of `IChaosInterceptor` upon client creation to inject faults. This PR will later be followed by a separate comprehensive Fault Injection library with a full implementation of the interface.

### PR DESCRIPTION
**Pull request by NaluTripician**
# Pull Request Template

## Description

Fault Injection Feature: #3819

This PR will add support for adding fault injection for the V3 .NET SDK. This will allow for adding an implementation of `IChaosInterceptor` upon client creation to inject faults. This PR will later be followed by a separate comprehensive Fault Injection library with a full implementation of the interface. 

Methods added to the `GlobalEndpointManager` will be used by the Fault Injection Library to filter rule applications by region endpoint.

Testing will be added for the implementation of all abstract classes in the Fault Injection Library. 

## Type of change

Please delete options that are not relevant.

- [] New feature (non-breaking change which adds functionality)

## Closing issues

To automatically close an issue: closes #4256

### CODE REVIEW

**Comment by NaluTripician** [comment]
Wait to merge PR until next version of direct package (3.32.1) is available as this PR is dependent on changes on thatversion.

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs [hunk 1]
```diff
@@ -483,6 +488,7 @@ internal partial class DocumentClient : IDisposable, IAuthorizationTokenProvider
             this.transportClientHandlerFactory = transportClientHandlerFactory;
             this.IsLocalQuorumConsistency = isLocalQuorumConsistency;
             this.initTaskCache = new AsyncCacheNonBlocking<string, bool>(cancellationToken: this.cancellationTokenSource.Token);
+            this.chaosInterceptor = chaosInterceptorFactory.CreateInterceptor(this);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/DocumentClient.cs:491]: This can throw a NullReferenceException. Because `chaosInterceptorFactory` can be null

Comment (NaluTripician) [reply Microsoft.Azure.Cosmos/src/DocumentClient.cs]: Added a null check.

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 2]
```diff
@@ -15,7 +15,8 @@ namespace Microsoft.Azure.Cosmos
     using System.Security.Cryptography.X509Certificates;
     using Microsoft.Azure.Cosmos.Fluent;
     using Microsoft.Azure.Documents;
-    using Microsoft.Azure.Documents.Client;
+    using Microsoft.Azure.Documents.Client;
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:18]: Check that all namespaces are needed. Is `using Microsoft.Azure.Documents.Client;` needed?

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs [hunk 3]
```diff
@@ -483,6 +488,7 @@ internal partial class DocumentClient : IDisposable, IAuthorizationTokenProvider
             this.transportClientHandlerFactory = transportClientHandlerFactory;
             this.IsLocalQuorumConsistency = isLocalQuorumConsistency;
             this.initTaskCache = new AsyncCacheNonBlocking<string, bool>(cancellationToken: this.cancellationTokenSource.Token);
+            this.chaosInterceptor = chaosInterceptorFactory == null ? chaosInterceptorFactory.CreateInterceptor(this) : null;
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/DocumentClient.cs:491]:
```suggestion
            this.chaosInterceptor = chaosInterceptorFactory?.CreateInterceptor(this);
```

File: Microsoft.Azure.Cosmos/src/Routing/GlobalEndpointManager.cs [hunk 4]
```diff
@@ -352,6 +352,26 @@ public virtual Uri ResolveServiceEndpoint(DocumentServiceRequest request)
         {
             return this.locationCache.ResolveServiceEndpoint(request);
         }
+
+        public Uri ResolveFaultInjectionServiceEndpoint(string region, bool writeOnly)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Routing/GlobalEndpointManager.cs:356]: THese new APIs, should they be added to `IGlobalEndpointManager`?

File: Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs [hunk 5]
```diff
@@ -349,6 +364,27 @@ public ReadOnlyCollection<Uri> GetApplicableEndpoints(DocumentServiceRequest req
             return new ReadOnlyCollection<Uri>(applicableEndpoints);
         }
 
+        public Uri ResolveFaultInjectionEndpoint(string region, bool writeOnly)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs:367]: What is particular about this logic for Fault Injection to be on the method name?

File: Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs [hunk 6]
```diff
@@ -349,6 +364,27 @@ public ReadOnlyCollection<Uri> GetApplicableEndpoints(DocumentServiceRequest req
...
+        {
+            RegionNameMapper regionNameMapper = new RegionNameMapper();
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs:369]: Why is RegionNameMapper needed? Is the `region` something coming from the user that we don't know if it's valid?

File: Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs [hunk 7]
```diff
@@ -349,6 +364,27 @@ public ReadOnlyCollection<Uri> GetApplicableEndpoints(DocumentServiceRequest req
...
+            if (writeOnly)
+            {
+                if (this.locationInfo.AvailableWriteEndpointByLocation.TryGetValue(regionNameMapper.GetCosmosDBRegionName(region), out Uri faultInjectionEndpoint))
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs:372]: If `AvailableWriteEndpointByLocation` is available externally (you are adding these public getters), why can't this logic be on the Fault Injection project?

File: Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs (hunk 5 above)
Comment (NaluTripician) [reply Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs:367]: This method just gets the region endpoint from the region name. This is only being used when creating a fault injection rule that targets a specific region endpoint. I am ok with changing the name. Does `GetRegionEndpointFromRegionName` work better?

File: Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs (hunk 6 above)
Comment (NaluTripician) [reply Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs:369]: Yes, we do not know if region is valid as this comes from the user. Looks like there were changes to master where a global `RegionNameMapper` is created for LocationCache, so switched to using that one.

File: Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs (hunk 7 above)
Comment (NaluTripician) [reply Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs:372]: `LocationInfo` which has `AvailableWriteEndpointByLocation` is private and not accessible so public getters are needed.

File: Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs (hunk 5 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs:364]: My question would be, if this is only for Fault Injection, cannot you have the code in the Fault Injection library then? It should have access to all these internal APIs

File: Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs (hunk 7 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs]: GlobalEndpointManager has the APIs to get the AvailableWriteEndpointByLocation list, right? So Fault Injection library which has internal access, can call the GlobalEndpointManager API, and the RegionMapper, and do this logic?

Comment (NaluTripician) [reply Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs]: Will remove this method and add logic to Fault Injection Library

File: Microsoft.Azure.Cosmos/src/Fluent/CosmosClientBuilder.cs [hunk 8]
```diff
@@ -658,6 +658,12 @@ internal CosmosClientBuilder WithPartitionLevelFailoverEnabled()
             return this;
         }
 
+        internal CosmosClientBuilder WithFaultInjection(IChaosInterceptorFactory chaosInterceptorFactory)
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/Fluent/CosmosClientBuilder.cs:661]: Can we add `<summary>` section here ?

File: Microsoft.Azure.Cosmos/src/FaultInjection/IChaosInterceptorFactory.cs [hunk 9]
```diff
@@ -0,0 +1,12 @@
+﻿//------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+//------------------------------------------------------------
+namespace Microsoft.Azure.Cosmos
+{
+    using Microsoft.Azure.Documents.FaultInjection;
+
+    internal interface IChaosInterceptorFactory
+    {
+        public IChaosInterceptor CreateInterceptor(DocumentClient documentClient);
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/FaultInjection/IChaosInterceptorFactory.cs:10]: Can we add `<summary>` section here ?

File: Microsoft.Azure.Cosmos/src/Routing/GlobalEndpointManager.cs [hunk 10]
```diff
@@ -351,7 +351,22 @@ private static bool IsNonRetriableException(Exception exception)
         public virtual Uri ResolveServiceEndpoint(DocumentServiceRequest request)
         {
             return this.locationCache.ResolveServiceEndpoint(request);
-        }
+        }
+
+        public Uri GetDefaultEndpoint()
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/Routing/GlobalEndpointManager.cs:356]: Please add more details in `<summary>` section around what each method is supposed to do.

**Review by kirankumarkolli — APPROVED** [review]

**Comment by kirankumarkolli** [comment]
Please update PR description on why API changes to the GEM and how it will be used.

**Pull request merged**